README.md
# App Server Chef Cookbook

Chef Cookbook for configuring and managing the Application Server within my Homelab Environment.

# Cookbook Goals:

### OS Customisation
* Disable SELinux
* Setup a system user (apps)
  - UID = 1550
  - GID = 1550
* Setup a personal user (cpressland)
  - UID = 1551
  - GID = 1551
  - Install SSH Key
  - Add to wheel group
  - Add to docker group
* Update Sudoers file to let members of the 'wheel' group run sudo without a password


### Local Applications
* Install and Configure Samba
  - Configure Samba for guess only access, map guest to 'apps' user.
  - Broadcast Downloads Share (/downloads)
* Install and Configure AutoFS
  - Replace default autofs.service systemd service file
    * Add "before=docker.service" to let AutoFS start before Docker
    * Add "ExecStartPost=" option to mount persistent storage immediately
  - Mount network share as 'apps' user
* Install and Configure Docker

### Permanent Docker Containers
* Deploy and Configure [cpressland/nginx](https://hub.docker.com/r/cpressland/nginx/) Docker Container
* Deploy and Configure [cpressland/php](https://hub.docker.com/r/cpressland/php/) Docker Container
* Deploy and Configure [docker-library/mariadb](https://hub.docker.com/_/mariadb/) Docker Container
* Deploy and Configure [docker-library/ghost](https://hub.docker.com/_/ghost/) Docker Container
* Deploy and Configure [cpressland/nzbget](https://hub.docker.com/r/cpressland/nzbget/) Docker Container
* Deploy and Configure [cpressland/sonarr](https://hub.docker.com/r/cpressland/sonarr/) Docker Container
* Deploy and Configure [cpressland/couchpotato](https://hub.docker.com/r/cpressland/couchpotato/) Docker Container
* Deploy and Configure [cpressland/plex](https://hub.docker.com/r/cpressland/plex/) Docker Container

### Transient Docker Containers
* Deploy and Configure [docker-library/ubuntu](https://hub.docker.com/_/ubuntu/) Docker Container
  - Used for restore of data from persistent storage to Docker Volumes
  - Used for backup of data from Docker Volumes to persistent storage

### Docker Volumes, Mount Points & Connections
* /downloads
  * /downloads
    * cpressland/nzbget
    * cpressland/sonarr
    * cpressland/couchpotato
* data_databases
  * /var/lib/mysql
    * docker-library/mariadb
* data_ghost
  * /var/lib/ghost
    * docker-library/ghost
* data_www
  * /var/www
    * cpressland/nginx
    * cpressland/php
* config_nginx
  * /etc/nginx
    * cpressland/nginx
* config_nzbget
  * /config
    * cpressland/nzbget
* config_sonarr
  * /config
    * cpressland/sonarr
* config_couchpotato
  * /config
    * cpressland/couchpotato
* config_plex
  * /config
    * cpressland/plex

## Supported Platforms

CentOS 7.2.1511

## Usage

Server can be bootstrapped with the following command:

```
curl -L https://raw.githubusercontent.com/cpressland/cookbook-appserver/master/install.sh | bash
```

Once the script has finished various Rake Tasks are available:

* `rake server:provision` - runs the 'default' recipe, essential
* `rake docker:restore` - Restores data to Docker Volumes from SMB Network Share
* `rake docker:backup` - Backs up data from Docker Volumes to SMB Network Share
* `rake docker:start` - Starts all configured Docker Containers
* `rake docker:stop` - Stops all chef configured Docker Containers
* `rake docker:destroy` - Stops and Removes all Docker Containers

## Author

Author:: Chris Pressland (mail@cpressland.io)


CHANGELOG.md
# 1.0.0

Initial release of appserver

# 1.0.1

Added Docker Network
Removed unused Firewall Ports
Removed Host Networking from Plex Server


attributes/default.rb

default['appserver']['packages'] = %w(wget tar git autofs samba samba-client cifs-utils curl htop vim firewalld)

default['storage']['share_ip'] = "10.0.50.10" # IP Address of unRAID Server
default['storage']['share_name'] = "shared" # Name of Share on unRAID Server

default['smb']['enable_smb'] = true
default['firewalld']['enable_firewalld'] = true

default['appserver']['repositories'] = [
  { :reponame=>"epel", :repodescription=>"Extra Packages for Enterprise Linux", :repobaseurl=>"http://mirrors.ukfast.co.uk/sites/dl.fedoraproject.org/pub/epel/7/x86_64/", :repogpgkey=>"http://dl.fedoraproject.org/pub/epel/RPM-GPG-KEY-EPEL-7"}
]

default['appserver']['users'] = [
  { :name=>"apps", :uid=>1550, :gid=>1550, :home=>"/home/apps" },
  { :name=>"cpressland", :uid=>1551, :gid=>1551, :home=>"/home/cpressland"}
]

default['appserver']['directories'] = [
  { :dirname=>"/downloads", :dirowner=>"apps", :dirgroup=>"apps", :dirmode=>"755" },
  { :dirname=>"/home/cpressland/.ssh", :dirowner=>"cpressland", :dirgroup=>"cpressland", :dirmode=>"700" }
]

default['appserver']['templates'] = [
 { :temname=>"/etc/auto.shared", :temsource=>"conf.auto.shared.erb", :temmode=>"644", :temowner=>"root", :temgroup=>"root"},
 { :temname=>"/etc/auto.master", :temsource=>"conf.auto.master.erb", :temmode=>"644", :temowner=>"root", :temgroup=>"root"},
 { :temname=>"/etc/samba/smb.conf", :temsource=>"conf.smb.conf.erb", :temmode=>"644", :temowner=>"root", :temgroup=>"root"},
 { :temname=>"/usr/lib/systemd/system/autofs.service", :temsource=>"systemd.autofs.service.erb", :temmode=>"644", :temowner=>"root", :temgroup=>"root"},
 { :temname=>"/home/cpressland/.ssh/authorized_keys", :temsource=>"conf.authorized_keys.erb", :temmode=>"600", :temowner=>"cpressland", :temgroup=>"cpressland"}
]

default['firewalld']['firewalld_ports'] = [
  { :fwport=>"32400/tcp", :fwzone=>"public" } # Plex
]

default['firewalld']['firewalld_services'] = [
  { :fwservice=>"ssh", :fwzone=>"public" }, # Local SSH
  { :fwservice=>"http", :fwzone=>"public" }, # Nginx
  { :fwservice=>"https", :fwzone=>"public" }, # Nginx
  { :fwservice=>"samba", :fwzone=>"public" } # Samba
]

default['docker']['images'] = [
  { :name=>"ubuntu", :tag=>"xenial" },
  { :name=>"cpressland/nginx", :tag=>"ubuntu" },
  { :name=>"cpressland/php", :tag=>"ubuntu" },
  { :name=>"mariadb", :tag=>"latest" },
  { :name=>"ghost", :tag=>"latest" },
  { :name=>"cpressland/nzbget", :tag=>"ubuntu" },
  { :name=>"cpressland/sonarr", :tag=>"ubuntu" },
  { :name=>"cpressland/couchpotato", :tag=>"ubuntu" },
  { :name=>"cpressland/plex", :tag=>"ubuntu" },
  { :name=>"cpressland/plexpy", :tag=>"ubuntu" },
  { :name=>"cpressland/netdata", :tag=>"ubuntu" }
]

default['docker']['volumes'] = %w(data_databases data_ghost data_www config_nginx config_netdata config_nzbget config_sonarr config_couchpotato config_plex config_plexpy)

default['docker']['restorecontainers'] = [
  { :name=>"restore-data_databases", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'data_databases:/data'], :command=>"/bin/tar xzf /docker/volumes/data_databases.tar.gz -C /data/ ." },
  { :name=>"restore-data_ghost", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'data_ghost:/data'], :command=>"/bin/tar xzf /docker/volumes/data_ghost.tar.gz -C /data/ ." },
  { :name=>"restore-data_www", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'data_www:/data'], :command=>"/bin/tar xzf /docker/volumes/data_www.tar.gz -C /data/ ." },
  { :name=>"restore-config_nginx", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_nginx:/data'], :command=>"/bin/tar xzf /docker/volumes/config_nginx.tar.gz -C /data/ ." },
  { :name=>"restore-config_nzbget", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_nzbget:/data'], :command=>"/bin/tar xzf /docker/volumes/config_nzbget.tar.gz -C /data/ ." },
  { :name=>"restore-config_sonarr", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_sonarr:/data'], :command=>"/bin/tar xzf /docker/volumes/config_sonarr.tar.gz -C /data/ ." },
  { :name=>"restore-config_couchpotato", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_couchpotato:/data'], :command=>"/bin/tar xzf /docker/volumes/config_couchpotato.tar.gz -C /data/ ." },
  { :name=>"restore-config_plex", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_plex:/data'], :command=>"/bin/tar xzf /docker/volumes/config_plex.tar.gz -C /data/ ." },
  { :name=>"restore-config_plexpy", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_plexpy:/data'], :command=>"/bin/tar xzf /docker/volumes/config_plexpy.tar.gz -C /data/ ." },
  { :name=>"restore-config_netdata", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_netdata:/data'], :command=>"/bin/tar xzf /docker/volumes/config_netdata.tar.gz -C /data/ ." }
]

default['docker']['permanentcontainers'] = [
  { :name=>"mariadb", :repo=>"mariadb", :tag=>"latest", :network_mode=>"cpressland.io", :volumes=>"data_databases:/var/lib/mysql" },
  { :name=>"php", :repo=>"cpressland/php", :tag=>"ubuntu", :network_mode=>"cpressland.io", :volumes=>"data_www:/var/www" },
  { :name=>"ghost", :repo=>"ghost", :tag=>"latest", :network_mode=>"cpressland.io", :volumes=>"data_ghost:/var/lib/ghost" },
  { :name=>"nzbget", :repo=>"cpressland/nzbget", :tag=>"ubuntu", :network_mode=>"cpressland.io", :volumes=>['config_nzbget:/config', '/downloads:/downloads'] },
  { :name=>"sonarr", :repo=>"cpressland/sonarr", :tag=>"ubuntu", :network_mode=>"cpressland.io", :volumes=>['config_sonarr:/config', '/downloads:/downloads', '/media/shared/px01/tv:/tv', '/dev/rtc:/dev/rtc:ro'] },
  { :name=>"couchpotato", :repo=>"cpressland/couchpotato", :tag=>"ubuntu", :network_mode=>"cpressland.io", :volumes=>['config_couchpotato:/config', '/downloads:/downloads', '/media/shared/px01/movies:/movies', '/etc/localtime:/etc/localtime:ro'] },
  { :name=>"plexpy", :repo=>"cpressland/plexpy", :tag=>"ubuntu", :network_mode=>"cpressland.io", :volumes=>"config_plexpy:/config" },
  { :name=>"plex", :repo=>"cpressland/plex", :tag=>"ubuntu", :network_mode=>"host", :volumes=>['config_plex:/config', '/media/shared/px01/tv:/tv', '/media/shared/px01/movies:/movies'] },
  { :name=>"nginx", :repo=>"cpressland/nginx", :tag=>"ubuntu", :network_mode=>"cpressland.io", :port=>['80:80', '443:443'], :volumes=>['data_www:/var/www', 'config_nginx:/etc/nginx'] },
  { :name=>"netdata", :repo=>"cpressland/netdata", :tag=>"ubuntu", :network_mode=>"cpressland.io", :cap_add=>"SYS_PTRACE", :volumes=>['/proc:/host/proc:ro', '/sys:/host/sys:ro', '/var/run/docker.sock:/var/run/docker.sock', 'config_netdata:/etc/netdata/'] }
]

default['docker']['backupcontainers'] = [
  { :name=>"backup-data_databases", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'data_databases:/data'], :command=>"/bin/tar czf /docker/volumes/data_databases.tar.gz -C /data/ ." },
  { :name=>"backup-data_ghost", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'data_ghost:/data'], :command=>"/bin/tar czf /docker/volumes/data_ghost.tar.gz -C /data/ ." },
  { :name=>"backup-data_www", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'data_www:/data'], :command=>"/bin/tar czf /docker/volumes/data_www.tar.gz -C /data/ ." },
  { :name=>"backup-config_nginx", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_nginx:/data'], :command=>"/bin/tar czf /docker/volumes/config_nginx.tar.gz -C /data/ ." },
  { :name=>"backup-config_nzbget", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_nzbget:/data'], :command=>"/bin/tar czf /docker/volumes/config_nzbget.tar.gz -C /data/ ." },
  { :name=>"backup-config_sonarr", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_sonarr:/data'], :command=>"/bin/tar czf /docker/volumes/config_sonarr.tar.gz -C /data/ ." },
  { :name=>"backup-config_couchpotato", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_couchpotato:/data'], :command=>"/bin/tar czf /docker/volumes/config_couchpotato.tar.gz -C /data/ ." },
  { :name=>"backup-config_plex", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_plex:/data'], :command=>"/bin/tar czf /docker/volumes/config_plex.tar.gz -C /data/ ." },
  { :name=>"backup-config_plexpy", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_plexpy:/data'], :command=>"/bin/tar czf /docker/volumes/config_plexpy.tar.gz -C /data/ ." },
  { :name=>"backup-config_netdata", :repo=>"ubuntu", :tag=>"xenial", :volumes=>['/media/shared/docker:/docker', 'config_netdata:/data'], :command=>"/bin/tar czf /docker/volumes/config_netdata.tar.gz -C /data/ ." }
]
